### PR TITLE
MAILBOX-374 Use a single channel for Event dispatching 

### PR DESCRIPTION
**Pull request by chibenwa**
Before, a channel was used for every call to `Sender::send`.

We now use an infinite flux on which we publish the events to use a single channel for this.

This is a temporary workaround, waiting for cleaner channel-pool management in next release of rabbitmq-reactor.

### CODE REVIEW

**Comment by chibenwa** [comment]
`AccessTokenRepositoryTest.validTokenMustBeRetrieved » Completion org.apache.ja...` test this please

**Comment by chibenwa** [comment]
This do solve the Gatling performance issue spotted (no more progress done after 2.000 events)

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 1]
```diff
@@ -67,6 +73,14 @@ void start() {
             .durable(DURABLE)
             .type(DIRECT_EXCHANGE))
             .block();
+
+        emitterProcessor = EmitterProcessor.create();
+        consumerSubscripotion = sender.send(emitterProcessor.publish().autoConnect()).subscribe();
```
Comment (trantienduchn) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: s/consumerSubscripotion/consumerSubscription

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java [hunk 2]
```diff
@@ -91,16 +105,22 @@ void start() {
...
-        Flux<OutboundMessage> outboundMessages = routingKeys
-            .flatMap(routingKey -> serializedEvent
-                .map(payload -> new OutboundMessage(MAILBOX_EVENT_EXCHANGE_NAME, routingKey.asString(), basicProperties, payload)));
+    Mono<Void> doDispatch(Mono<byte[]> serializedEvent, String exchangeName, Stream<RoutingKeyConverter.RoutingKey> routingKeyStream, AMQP.BasicProperties properties) {
+        Stream<OutboundMessage> outboundMessages = routingKeyStream
+            .map(routingKey -> new OutboundMessage(exchangeName, routingKey.asString(), properties, serializedEvent.block()));
 
-        return sender.send(outboundMessages);
+        return Mono.fromRunnable(() -> {
+            synchronized (lock) {
```
Comment (trantienduchn) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: why did you need to lock? I think any reactor object is suppose to work well under concurrency environtment

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java [hunk 3]
```diff
@@ -41,6 +43,26 @@ void ofShouldDeserializeUUIDs() {
         assertThat(EventBusId.of(UUID_1.toString()))
             .isEqualTo(EventBusId.of(UUID_1));
     }
+    @Test
+    void oo() {
```
Comment (trantienduchn) [inline mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java]: `oo` ?

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: `I think any reactor object is suppose to work well under concurrency environtment`

This assumption is wrong: the producer needs to be on a single thread to cope with Rx specifications.

Please have a look to FluxProcessor subclass javadoc to get a glance of the difficulty of that topic.

Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: A lock ensures things go fine.

File: mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java (hunk 3 above)
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/test/java/org/apache/james/mailbox/events/EventBusIdTest.java]: Debugging statement that got lost. Sorry.

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java (hunk 2 above)
Comment (trantienduchn) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/EventDispatcher.java]: what is different if I created a method likes below and use it in Mono.fromRunnable()? I think the code will look better. Anw, that is just a tiny comment.
```java
sychronized emit(OutboundMessage message) {
    emmiterProcessor.onNext(message);
}
```

**Comment by chibenwa** [comment]
Duplicates https://github.com/linagora/james-project/pull/2157

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/2157
